js/pro/krakenfutures.js

```diff
@@ -0,0 +1,475 @@
+        //     trades: [
+        //       {
+        //         feed: 'trade',
+        //         product_id: 'PF_XBTUSD',
+        //         uid: 'e074476f-09de-4c39-9538-1c1fc485a3af',
+        //         side: 'buy',
+        //         type: 'fill',
+        //         seq: 46590,
+        //         time: 1677863497310,
+        //         qty: 1,
+        //         price: 22437
+        //       }
+        //     ]
+        //   }
+        //
+        // single trade
+        //
+        //   {
+        //       feed: 'trade',
+        //       product_id: 'PF_XBTUSD',
+        //       uid: 'cf0932f3-a048-4f23-821f-1cc24a50b922',
+        //       side: 'buy',
+        //       type: 'fill',
+        //       seq: 46690,
+        //       time: 1677865802580,
+        //       qty: 0.0489,
+        //       price: 22429
+        //   }
+        //
+        const trades = this.safeValue (message, 'trades', [ message ]);
+        const marketId = this.safeStringLower (message, 'product_id');
+        const market = this.safeMarket (marketId);
+        const symbol = market['symbol'];
+        let stored = this.safeValue (this.trades, symbol);
+        if (stored === undefined) {
+            const limit = this.safeInteger (this.options, 'tradesLimit', 1000);
+            stored = new ArrayCache (limit);
+            this.trades[symbol] = stored;
+        }
+        for (let j = 0; j < trades.length; j++) {
+            const parsed = this.parseTrade (trades[j], market);
+            stored.append (parsed);
+        }
+        const messageHash = 'trade' + ':' + symbol;
+        client.resolve (stored, messageHash);
+    }
+
+    handleSystemStatus (client, message) {
+        //
+        // todo: answer the question whether handleSystemStatus should be renamed
+        // and unified as handleStatus for any usage pattern that
+        // involves system status and maintenance updates
+        //
+        //     {
+        //         connectionID: 15527282728335292000,
+        //         event: 'systemStatus',
+        //         status: 'online', // online|maintenance|(custom status tbd)
+        //         version: '0.2.0'
+        //     }
+        //
+        return message;
+    }
+
+    async challenge () {
+        //
+        // Sample Sent Message
+        // {
+        //  "event": "challenge",
+        //  "api_key": "CMl2SeSn09Tz+2tWuzPfdaJdsahq6qv5UaexXuQ3SnahDQU/gO3aT+"
+        // }
+        //
+        const messageHash = 'challenge';
+        const url = this.urls['api']['ws'];
+        const client = this.client (url);
+        const challenge = this.safeString (client.subscriptions, messageHash);
+        if (challenge !== undefined) {
+            return challenge;
+        }
+        const request = {
+            'event': 'challenge',
+            'api_key': this.apiKey,
+        };
+        const future = client.future (messageHash);
+        await client.send (request);
+        await future;
```

Comment:
@carlosmiei I don't feel like doing correct way here. Do you have advices?

Reply:
@youknowone I will take a look soon and let you know 🙂